fence/man/fence_tool.8
.TH FENCE_TOOL 8 2009-12-21 cluster cluster

.SH NAME
fence_tool \- a utility for the fenced daemon

.SH SYNOPSIS
.B fence_tool
[COMMAND] [OPTIONS]

.SH DESCRIPTION
This utility controls and queries the
.BR fenced (8)
daemon with the following commands:
.TP
.B join
join the fence domain.
.TP
.B leave
leave the fence domain.
.TP
.B dump
print the fenced internal debug buffer ont stdout.
.TP
.B ls
display internal fenced state.
.P
The leave command will not be sent to fenced if fence_tool detects that
any instances of gfs or dlm are in use.

.SH OPTIONS
.TP
.B \-n
Show all node information in ls.
.TP
.BI \-t " seconds"
Retry cman connection for this many seconds.
0 none, -1 indefinite. Default 0.
.TP
.BI \-q " seconds"
Delay join up to this many seconds for the cluster to have quorum.
0 none, -1 indefinite. Default 0.
.TP
.BI \-m " seconds"
Delay join up to this many seconds for all nodes in cluster.conf to be
cluster members.
0 none, -1 indefinite. Default 0.
.TP
.BI \-w " seconds"
Wait up to this many seconds for the result of join or leave.
0 none, -1 indefinite. Default 0.
.TP
.B \-h
Print a help message describing available options, then exit.
.TP
.B \-V
Print program version information, then exit.

.SH SEE ALSO
.BR fenced (8)



rgmanager/ChangeLog
2008-04-16 Fabio M. Di Nitto <fabbione at fabbione.net>
	* src/utils/clustat.c: Fix variable init

2008-03-04 Lon Hohberger <lhh at redhat.com>
	* src/resources/ASEHAagent.sh, Makefile: Add Sybase failover agent
	* src/resources/oracledb.sh, Makefile: Add Oracle 10g failover agent

2008-02-26 Lon Hohberger <lhh at redhat.com>
	* src/resources/ip.sh: Fix netmask handling in ip.sh
	* src/utils/clustat.c: Don't show estranged nodes if they're down
2008-02-01 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_state.c, slang_event.c, groups.c, restart_counter.c,
	include/restart_counter.h, resgroup.h,
	src/resources/default_event_script.sl: Allow restart counters to 
	correctly work with central_processing. (#400211 / #431130)

2008-01-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_thread.c: Fix case that broke 'clusvcadm -e <service>
	-n <node>' case #430220
	* src/daemons/rg_state.c: Don't enable migrate-to-self (#430272)

2008-01-02 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/resources/ocf-shellfuncs: Add interpreter.

2007-12-30 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/daemons/rg_state.c, include/resgroup.h, src/clulib/alloc.c:
	  Fix building when -DDEBUG is defined.

2007-12-24 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/clulib/vft.c: Change ifdef to fix build on parisc.

2007-12-19 Lon Hohberger <lhh at redhat.com>
	* include/res-ocf.h: Fix typo
	* src/daemons/restree.c: Export reference counts to resource agents.
	* src/resources/clusterfs.sh: Use reference counts to ensure we
	don't unmount a GFS file system if it's in use by multiple services
	on the same node.  #254111
	* src/resources/service.sh, default_event_script.sl: Add support for
	"soft" dependencies if central_processing is enabled

2007-12-14 Lon Hohberger <lhh at redhat.com>
	(Misc central_events fixes)
	* Add return code for inability to run because of exclusive 
	tag
	* Only initialize state once
	* Fix master-node failover.
	* Correctly return node after enable if request was forwarded

2007-12-12 Lon Hohberger <lhh at redhat.com>
	* Misc changes; add missing ds.h
	* src/resources/default*.sl: Make clusvcadm -r go to a different
	node.  Don't notice() that we're starting a service if it's
	disabled.
	* src/daemons/groups.c: Make clustat not interfere with startup
	when central processing is enabled
	* src/daemons/slang_event.c: Make clusvcadm correctly report
	service owner after relocate/restart/enable requests.

2007-11-30 Lon Hohberger <lhh at redhat.com>
	* Commit RIND / S-Lang script engine [untested]

[RHEL5 merged ChangeLog Entries]
2007-11-30 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Retry mount up to 3 times to avoid
	race condition during another process mounting a GFS volume
	* src/resources/vm.sh, service.sh: Add defaults for values.
	Make vm.sh work with more service attrs (max restarts)
	* src/utils/clustat.c: Make output of clustat terminal-width 
	dependent

2007-11-26 Lon Hohberger <lhh at redhat.com>
	* include/reslist.h: Add restart counters to resource node structure
	(intended for top-level resources, i.e. services, vms...)
	* include/restart_counter.h: Add header file for restart counter
	* src/daemons/Makefile: Fix build to include restart counters
	* src/daemons/restart_counter.c: Implement restart counters #247139
	* src/daemons/fo_domain.c, groups.c, restart_counter.c, resrules.c,
	restree.c, test.c: Glue for restart counters.
	* src/daemons/reslist.c: Glue for restart counters.  Make expand_time
	parser more robust to allow things like '1h30m' as a time value.
	* src/daemons/main.c: Mark quorum disk offline in the correct
	place to avoid extraneous log messages
	* src/daemons/rg_state.c: Allow marking service as stopped if
	stuck in recover state.  Make service which failed to start
	go to stopped state.  Glue for restart counters.
	* src/resources/service.sh, vm.sh: Add parameters for restart
	counters #247139

2007-11-14 Lon Hohberger <lhh at redhat.com>
	* src/utils/clulog.c: Make clulog honor rgmanager log levels
	(#289501)
	* src/clulib/vft.c: Fix #303981 - crash on rgmanager restart in some
	cases
	* man/clusvcadm.8: Remove references to clushutdown from man page;
	resolves #324151
	* src/resources/netfs.sh: Apply patch from Marco Ceci to fix #358161
	* src/resources/vm.sh: Make default migration policy live instead
	of pause for Xen virtual machines.  Also make it configurable instead
	of static.  Resolves #345871

2007-11-13 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Add support for self_fence operation
	to clusterfs resource agent
	* src/resources/service.sh: Add default values to service.sh

2007-10-26 Lon Hohberger <lhh at redhat.com>
	* src/daemons/main.c, src/utils/clustat.c, clusvcadm.c:
	Fix #354391

2007-09-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c: Apply patch to fix side case re: 229650
	Patch from Simone Gotti. Resolves: #229650
[End RHEL5 merged changes]

2007-08-30 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c, rg_state.c: Fix tree-restart bug
	This is another part of #229650

2007-08-22 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* include/list.h: Fix build with gcc-4.2.

2007-08-15 Lon Hohberger <lhh at redhat.com>
	* src/utils/clustat.c: Remove uninitialized var & fix logic /
	spotted by Ryan McCabe

2007-08-09  Fabio M. Di Nitto <fabbione at ubuntu.com>
	* rgmanager/src/clulib/alloc.c: Undefine DEBUG when building on
	PARISC. __builtin_address needs the same love as IA64.

2007-08-02 Lon Hohberger <lhh at redhat.com>
	* general: More fixes around #248727
	* include/reslist.h, src/daemons/restree.c: Make last-value be
	returned or resources which have been checked recently
	* src/daemons/groups.c: Make VMs use migrate semantics instead of
	relocate semantics when employing failover domain rules
	* src/daemons/nodeevent.c: Fix VMs ending up on wrong nodes when
	simultaneous boot occurs
	* src/daemons/rg_forward.c: Fix erroneous timeout

fence/fence_tool/fence_tool.c
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <signal.h>
#include <string.h>
#include <stdint.h>
#include <dirent.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <fcntl.h>
#include <limits.h>
#include <errno.h>
#include <libgen.h>

#include "ccs.h"
#include "copyright.cf"
#include "libcman.h"
#include "libfenced.h"

#define OP_JOIN  			1
#define OP_LEAVE 			2
#define OP_LIST				3
#define OP_DUMP				4

#define MAX_NODES			128

int all_nodeids[MAX_NODES];
int all_nodeids_count;
cman_handle_t ch;
cman_node_t cman_nodes[MAX_NODES];
int cman_nodes_count;
struct fenced_node nodes[MAX_NODES];
char *prog_name;
int operation;

#define DEFAULT_RETRY_CMAN 0 /* fail immediately if we can't connect to cman */
#define DEFAULT_DELAY_QUORUM 0
#define DEFAULT_DELAY_MEMBERS 0
#define DEFAULT_WAIT_JOINLEAVE 0

int opt_all_nodes = 0;
int opt_retry_cman = DEFAULT_RETRY_CMAN;
int opt_delay_quorum = DEFAULT_DELAY_QUORUM;
int opt_delay_members = DEFAULT_DELAY_MEMBERS;
int opt_wait_joinleave = DEFAULT_WAIT_JOINLEAVE;


#define die(fmt, args...) \
do { \
	fprintf(stderr, "%s: ", prog_name); \
	fprintf(stderr, fmt "\n", ##args); \
	exit(EXIT_FAILURE); \
} while (0)

static int do_write(int fd, void *buf, size_t count)
{
	int rv, off = 0;

 retry:
	rv = write(fd, (char *)buf + off, count);
	if (rv == -1 && errno == EINTR)
		goto retry;
	if (rv < 0)
		return rv;

	if (rv != count) {
		count -= rv;
		off += rv;
		goto retry;
	}
	return 0;
}

#define LOCKFILE_NAME           "/var/run/fenced.pid"

static void check_fenced_running(void)
{
	struct flock lock;
	int fd, rv;

	fd = open(LOCKFILE_NAME, O_RDONLY);
	if (fd < 0)
		die("fenced not running, no lockfile");

	lock.l_type = F_RDLCK;
	lock.l_start = 0;
	lock.l_whence= SEEK_SET;
	lock.l_len = 0;

	rv = fcntl(fd, F_GETLK, &lock);
	if (rv < 0)
		die("fenced not running, get lockfile");

	if (lock.l_type == F_UNLCK)
		die("fenced not running, unlocked lockfile");

	close(fd);
}

static int check_gfs(void)
{
	FILE *file;
	char line[PATH_MAX];
	char device[PATH_MAX];
	char path[PATH_MAX];
	char type[PATH_MAX];
	int count = 0;

	file = fopen("/proc/mounts", "r");
	if (!file)
		return 0;

	while (fgets(line, PATH_MAX, file)) {
		if (sscanf(line, "%s %s %s", device, path, type) != 3)
			continue;
		if (!strcmp(type, "gfs") || !strcmp(type, "gfs2")) {
			fprintf(stderr, "found %s file system mounted from %s on %s\n",
				type, device, path);
			count++;
		}
	}

	fclose(file);
	return count;
}

static int check_controlled_dir(const char *path)
{
	DIR *d;
	struct dirent *de;
	int count = 0;

	d = opendir(path);
	if (!d)
		return 0;

	while ((de = readdir(d))) {
		if (de->d_name[0] == '.')
			continue;

#if 0
		if (strstr(path, "fs/gfs") && ignore_nolock(path, de->d_name))
			continue;
#endif

		fprintf(stderr, "found dlm lockspace %s/%s\n", path, de->d_name);
		count++;
	}

	closedir(d);
	return count;
}

/* Copying fenced's check_uncontrolled_entries()/check_controlled_dir()
   in part here.
   We could also use check_controlled_dir to detect gfs file systems intead
   of looking at /proc/mounts... /proc/mounts gives us more info (mountpoint)
   to report about the offending fs */

static void check_controlled_systems(void)
{
	int count = 0;

	count += check_gfs();
	count += check_controlled_dir("/sys/kernel/dlm");

	if (count)
		die("cannot leave due to active systems");
}

static int we_are_in_fence_domain(void)
{
	struct fenced_node nodeinfo;
	int rv;

	memset(&nodeinfo, 0, sizeof(nodeinfo));

	rv = fenced_node_info(FENCED_NODEID_US, &nodeinfo);
	if (rv < 0)
		return 0;

	if (nodeinfo.member)
		return 1;
	return 0;
}

static void wait_domain(int joining)
{
	int in, tries = 0;

	while (1) {
		if (joining)
			check_fenced_running();

		in = we_are_in_fence_domain();

		if (joining && in)
			break;

		if (!joining && !in)
			break;

		tries++;

		if (opt_wait_joinleave < 0)
			goto retry_domain;

		if (!opt_wait_joinleave || tries >= opt_wait_joinleave) {
			fprintf(stderr, "%s: %s not complete\n",
			       prog_name, joining ? "join" : "leave");
			break;
		}
 retry_domain:
		if (!(tries % 10))
			fprintf(stderr, "%s: waiting for fenced to %s the fence group.\n",
			       prog_name, joining ? "join" : "leave");

		sleep(1);
	}

	return;
}

static void read_ccs_nodeids(int cd)
{
	char path[PATH_MAX];
	char *nodeid_str;
	int i, error;

	memset(all_nodeids, 0, sizeof(all_nodeids));
	all_nodeids_count = 0;

	for (i = 1; ; i++) {
		nodeid_str = NULL;
		memset(path, 0, sizeof(path));
		sprintf(path, "/cluster/clusternodes/clusternode[%d]/@nodeid", i);

		error = ccs_get(cd, path, &nodeid_str);
		if (error || !nodeid_str)
			break;

		all_nodeids[all_nodeids_count++] = atoi(nodeid_str);
		free(nodeid_str);
	}
}

static int all_nodeids_are_members(void)
{
	int i, j, rv, found;

	memset(&cman_nodes, 0, sizeof(cman_nodes));
	cman_nodes_count = 0;

	rv = cman_get_nodes(ch, MAX_NODES, &cman_nodes_count, cman_nodes);
	if (rv < 0)
		return -1;

	for (i = 0; i < all_nodeids_count; i++) {
		found = 0;

		for (j = 0; j < cman_nodes_count; j++) {
			if (cman_nodes[j].cn_nodeid == all_nodeids[i] &&
			    cman_nodes[j].cn_member) {
				found = 1;
				break;
			}
		}

		if (!found)
			return 0;
	}
	return 1;
}

static int connect_cman(void)
{
	int rv, tries = 0;

	while (1) {
		ch = cman_init(NULL);
		if (ch)
			break;

		tries++;

		if (opt_retry_cman < 0)
			goto retry_init;

		if (!opt_retry_cman || tries >= opt_retry_cman)
			return -1;
 retry_init:
		if (!(tries % 10))
			fprintf(stderr, "%s: retrying cman connection\n", prog_name);
		sleep(1);
	}

	while (1) {
		rv = cman_is_active(ch);
		if (rv)
			break;

		tries++;

		if (opt_retry_cman < 0)
			goto retry_active;

		if (!opt_retry_cman || tries >= opt_retry_cman) {
			cman_finish(ch);
			return -1;
		}
 retry_active:
		if (!(tries % 10))
			fprintf(stderr, "%s: retrying cman active check\n", prog_name);
		sleep(1);
	}

	return 0;
}

static void delay_quorum(void)
{
	int rv, tries = 0;

	while (1) {
		rv = cman_is_quorate(ch);
		if (rv)
			break;

		rv = cman_is_active(ch);
		if (!rv) {
			cman_finish(ch);
			die("lost cman connection");
		}
		
		tries++;

		if (opt_delay_quorum < 0)
			goto retry_quorum;

		if (!opt_delay_quorum || tries >= opt_delay_quorum) {
			fprintf(stderr, "%s: continuing without quorum\n", prog_name);
			break;
		}
 retry_quorum:
		if (!(tries % 10))
			fprintf(stderr, "%s: delaying for quorum\n", prog_name);

		sleep(1);
	}

	return;
}

static void delay_members(void)
{
	int rv, tries = 0;
	int cd;

	cd = ccs_connect();
	if (cd < 0) {
		cman_finish(ch);
		die("lost cman/ccs connection");
	}

	read_ccs_nodeids(cd);

	while (1) {
		rv = all_nodeids_are_members();
		if (rv < 0) {
			ccs_disconnect(cd);
			cman_finish(ch);
			die("lost cman connection");
		}
		if (rv)
			break;

		tries++;

		if (opt_delay_members < 0)
			goto retry_members;

		if (!opt_delay_members || tries > opt_delay_members) {
			fprintf(stderr, "%s: continuing without all members\n", prog_name);
			break;
		}
 retry_members:
		if (!(tries % 10))
			fprintf(stderr, "%s: delaying for members\n", prog_name);

		sleep(1);
	}

	ccs_disconnect(cd);
	return;
}

static void do_join(int argc, char *argv[])
{
	int rv, tries = 0;

	rv = connect_cman();
	if (rv < 0)
		die("can't connect to cman");

	/* if delay_quorum() or delay_members() fail on any cman/ccs
	   connection or operation, they call cman_finish() and exit
	   with failure */

	if (opt_delay_quorum)
		delay_quorum();

	if (opt_delay_members)
		delay_members();

	cman_finish(ch);

	/* This loop deals with the case where fenced is slow enough starting
	   up that fenced_join fails.  Do we also want to add a delay here to
	   deal with the case where fenced is so slow starting up that it hasn't
	   locked its lockfile yet, causing check_fenced_running to fail? */

	while (1) {
		rv = fenced_join();
		if (!rv)
			break;

		check_fenced_running();

		tries++;
		if (!(tries % 10))
			fprintf(stderr, "%s: retrying join\n", prog_name);
		sleep(1);
	}

	if (opt_wait_joinleave)
		wait_domain(1);

	exit(EXIT_SUCCESS);
}

static void do_leave(void)
{
	int rv;

	check_controlled_systems();

	rv = fenced_leave();
	if (rv < 0)
		die("leave: can't communicate with fenced");

	if (opt_wait_joinleave)
		wait_domain(0);

	exit(EXIT_SUCCESS);
}

static void do_dump(void)
{
	char buf[FENCED_DUMP_SIZE];
	int rv;

	rv = fenced_dump_debug(buf);
	if (rv < 0)
		die("dump: can't communicate with fenced");

	do_write(STDOUT_FILENO, buf, strlen(buf));

	exit(EXIT_SUCCESS);
}

static int node_compare(const void *va, const void *vb)
{
	const struct fenced_node *a = va;
	const struct fenced_node *b = vb;

	return a->nodeid - b->nodeid;
}

/* copied from fence/fenced/fd.h, should probably be in libfenced.h */
#define CGST_WAIT_CONDITIONS	1
#define CGST_WAIT_MESSAGES	2
#define CGST_WAIT_FENCING	3

static const char *wait_str(int state)
{
	switch (state) {
	case 0:
		return "none";
	case CGST_WAIT_CONDITIONS:
		return "quorum";
	case CGST_WAIT_MESSAGES:
		return "messages";
	case CGST_WAIT_FENCING:
		return "fencing";
	}
	return "unknown";
}

/* copied from fence/fenced/fd.h, should probably be in libfenced.h */
#define VIC_DONE_AGENT          1
#define VIC_DONE_MEMBER         2
#define VIC_DONE_OVERRIDE       3
#define VIC_DONE_EXTERNAL       4

static const char *how_str(int how)
{
	switch (how) {
	case 0:
		return "none";
	case VIC_DONE_AGENT:
		return "agent";
	case VIC_DONE_MEMBER:
		return "member";
	case VIC_DONE_OVERRIDE:
		return "override";
	case VIC_DONE_EXTERNAL:
		return "external";
	}
	return "unknown";
}

static int do_list(void)
{
	struct fenced_domain d;
	struct fenced_node *np;
	int node_count;
	int rv, i;

	rv = fenced_domain_info(&d);
	if (rv < 0)
		exit(EXIT_FAILURE); /* fenced probably not running */

	printf("fence domain\n");
	printf("member count  %d\n", d.member_count);
	printf("victim count  %d\n", d.victim_count);
	printf("victim now    %d\n", d.current_victim);
	printf("master nodeid %d\n", d.master_nodeid);
	printf("wait state    %s\n", wait_str(d.state));
	printf("members       ");

	node_count = 0;
	memset(&nodes, 0, sizeof(nodes));

	rv = fenced_domain_nodes(FENCED_NODES_MEMBERS, MAX_NODES,
				 &node_count, nodes);
	if (rv < 0) {
		printf("error\n");
		goto fail;
	}

	qsort(&nodes, node_count, sizeof(struct fenced_node), node_compare);

	np = nodes;
	for (i = 0; i < node_count; i++) {
		printf("%d ", np->nodeid);
		np++;
	}
	printf("\n");

	if (!opt_all_nodes) {
		printf("\n");
		exit(EXIT_SUCCESS);
	}

	node_count = 0;
	memset(&nodes, 0, sizeof(nodes));

	rv = fenced_domain_nodes(FENCED_NODES_ALL, MAX_NODES,
				 &node_count, nodes);
	if (rv < 0)
		goto fail;

	qsort(&nodes, node_count, sizeof(struct fenced_node), node_compare);

	printf("all nodes\n");

	np = nodes;
	for (i = 0; i < node_count; i++) {
		printf("nodeid %d member %d victim %d last fence master %d how %s\n",
				np->nodeid,
				np->member,
				np->victim,
				np->last_fenced_master,
				how_str(np->last_fenced_how));
		np++;
	}
	printf("\n");
	exit(EXIT_SUCCESS);
 fail:
	fprintf(stderr, "fenced query error %d\n", rv);
	printf("\n");
	exit(EXIT_FAILURE);
}

static void print_usage(void)
{
	printf("Usage:\n");
	printf("\n");
	printf("%s <ls|join|leave|dump> [options]\n", prog_name);
	printf("\n");
	printf("Actions:\n");
	printf("  ls		   List nodes status\n");
	printf("  join             Join the default fence domain\n");
	printf("  leave            Leave default fence domain\n");
	printf("  dump		   Dump debug buffer from fenced\n");
	printf("\n");
	printf("Options:\n");
	printf("  -n               Show all node information in ls\n");

	printf("  -t <seconds>     Retry cman connection for <seconds>.\n");
	printf("                   Default %d.  0 no retry, -1 indefinite retry.\n",
				   DEFAULT_RETRY_CMAN);

	printf("  -q <seconds>     Delay join up to <seconds> for the cluster to have quorum.\n");
	printf("                   Default %d.  0 no delay, -1 indefinite delay.\n",
				   DEFAULT_DELAY_QUORUM);

	printf("  -m <seconds>     Delay join up to <seconds> for all nodes in cluster.conf\n");
	printf("                   to be cluster members.\n");
	printf("                   Default %d.  0 no delay, -1 indefinite delay\n",
				   DEFAULT_DELAY_MEMBERS);

	printf("  -w <seconds>     Wait up to <seconds> for join or leave result.\n");
	printf("                   Default %d.  0 no wait, -1 indefinite wait.\n",
				   DEFAULT_WAIT_JOINLEAVE);

	printf("  -V               Print program version information, then exit\n");
	printf("  -h               Print this help, then exit\n");
	printf("\n");
}

#define OPTION_STRING "nt:q:m:w:Vh"

static void decode_arguments(int argc, char *argv[])
{
	int cont = 1;
	int optchar;

	while (cont) {
		optchar = getopt(argc, argv, OPTION_STRING);

		switch (optchar) {

		case 'n':
			opt_all_nodes = 1;
			break;

		case 't':
			opt_retry_cman = atoi(optarg);
			break;

		case 'q':
			opt_delay_quorum = atoi(optarg);
			break;

		case 'm':
			opt_delay_members = atoi(optarg);
			break;

		case 'w':
			opt_wait_joinleave = atoi(optarg);
			break;

		case 'V':
			printf("fence_tool %s (built %s %s)\n",
			       RELEASE_VERSION, __DATE__, __TIME__);
			printf("%s\n", REDHAT_COPYRIGHT);
			exit(EXIT_SUCCESS);
			break;

		case 'h':
			print_usage();
			exit(EXIT_SUCCESS);
			break;

		case ':':
		case '?':
			fprintf(stderr, "Please use '-h' for usage.\n");
			exit(EXIT_FAILURE);
			break;

		case EOF:
			cont = 0;
			break;

		default:
			die("unknown option: %c\n", optchar);
			break;
		}
	}

	while (optind < argc) {
		if (strcmp(argv[optind], "join") == 0) {
			operation = OP_JOIN;
		} else if (strcmp(argv[optind], "leave") == 0) {
			operation = OP_LEAVE;
		} else if (strcmp(argv[optind], "dump") == 0) {
			operation = OP_DUMP;
		} else if (strcmp(argv[optind], "ls") == 0) {
			operation = OP_LIST;
		} else
			die("unknown option %s\n", argv[optind]);
		optind++;
	}

	if (!operation)
		die("no operation specified\n");
}

int main(int argc, char *argv[])
{
	prog_name = basename(argv[0]);

	decode_arguments(argc, argv);

	switch (operation) {
	case OP_JOIN:
		do_join(argc, argv);
		break;
	case OP_LEAVE:
		do_leave();
		break;
	case OP_DUMP:
		do_dump();
		break;
	case OP_LIST:
		do_list();
		break;
	}

	return EXIT_FAILURE;
}

